### PR TITLE
reimplement `/data_objects/study/{study_id}` using alldocs

### LINKED ISSUE #401: Mock up an API endpoint that, given a `Study.id` and `DataObject.data_object_type`, returns the relevant `DataObject`s
This Issue (task) came out of today's metadata squad meeting (Wednesday, November 22, 2023).

---

Based on the first bullet point in https://github.com/microbiomedata/nmdc-runtime/issues/355...

...I'm envisioning an API endpoint that accepts a `Study.id` value and a `DataObject.data_object_type` value, and returns (i.e. responds with) a JSON array of all the `DataObjects` that are associated with that study and have that specific `data_object_type` value.

`DataObject` documentation:
- https://microbiomedata.github.io/nmdc-schema/DataObject/

Here's an example of a "long" query team members have run; not necessarily related to this Issue other than to serve as an example of a long query, which is something endpoints like this one could save people from having to write themselves:
- https://github.com/microbiomedata/notebook_hackathons/blob/soil-contig-tax/taxonomic_dist_by_soil_layer/python/mongodb_query.txt.js

### PR DESCRIPTION
**Pull request by sujaypatil96**
# Description

This PR seeks to reimplement the `/data_objects/study/{study_id}` endpoint that is currently part of the runtime API endpoints suite so that it accounts for the schema design considerations noted here: https://docs.google.com/document/d/1HglWENf1m9IDAuk6468OAFdCRvpC3wCQG8wwDHwll1M/edit?usp=sharing

Another point to note is that is that it also uses the `alldocs` collection for its implementation.

Fixes #401 

TODO: write tests for this API endpoint.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration, if it is not simply `make up-test && make test-run`.

# Definition of Done (DoD) Checklist:
A simple checklist of necessary activities that add verifiable/demonstrable value to the product by asserting the quality of a feature, not the functionality of that feature; the latter should be stated as acceptance criteria in the issue this PR closes. Not all activities in the PR template will be applicable to each feature since the definition of done is intended to be a comprehensive checklist. Consciously decide the applicability of value-added activities on a feature-by-feature basis.

- [x] My code follows the style guidelines of this project (have you run `black nmdc_runtime/`?)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (in `docs/` and in <https://github.com/microbiomedata/NMDC_documentation/>?)
- [ ] I have added tests that prove my fix is effective or that my feature works, incl. considering downstream usage (e.g. <https://github.com/microbiomedata/notebook_hackathons>) if applicable.
- [ ] New and existing unit and functional tests pass locally with my changes (`make up-test && make test-run`)

### CODE REVIEW

**Comment by sujaypatil96** [comment]
@PeopleMakeCulture i'm having a little trouble writing a test in `tests/test_api/test_endpoints.py` because I seem to be having some trouble running the tests that I'm writing. Would you be able to help me write a test for this sometime? Perhaps after the roundtable updates at squad sync call tomorrow?

**Comment by dwinston** [comment]
@sujaypatil96 One pattern for running tests that I've pushed in order to demo here:

```
# Ensure any dev or test stacks are down: `make down-dev`, `make down-test`
make up-test
make test-build

# Expand `make test-run` to run specific test only, using pytest `-k` flag.
docker compose --file docker-compose.test.yml run test -k 'test_find_data_objects_for_study'
```

**Comment by dwinston** [comment]
I was thinking you could implement as an aggregation pipeline with a `$graphLookup` stage. The following, for example, yields for me 4,395 documents based on the prod db backup from `2024-07-29_20-12-07`:

```js
db.getCollection("alldocs").aggregate(

    // Pipeline
    [
        // Stage 1
        {
            $match: {
                "type": "Biosample",
                "part_of": "nmdc:sty-11-hdd4bf83"
            }
        },

        // Stage 2
        {
            $project: {
                "biosample_id": "$id"
            }
        },

        // Stage 3
        {
            $graphLookup: {
                from: "alldocs",
                startWith: "$biosample_id",
                connectFromField: "has_output",
                connectToField: "has_input",
                as: "process"
            }
        },

        // Stage 4
        {
            $unwind: {
                path: "$process"
            }
        },

        // Stage 5
        {
            $match: {
                "process.has_output.0": {"$exists": true}
            }
        },

        // Stage 6
        {
            $unwind: {
                path: "$process.has_output"
            }
        },

        // Stage 7
        {
            $project: {
                "has_output": "$process.has_output",
                "biosample_id": 1
            }
        },

        // Stage 8
        {
            $lookup: {
                from: "data_object_set",
                localField: "has_output",
                foreignField: "id",
                as: "data_object"
            }
        },

        // Stage 9
        {
            $match: {
                "data_object.0": {"$exists": true}
            }
        },

        // Stage 10
        {
            $project: {
                "biosample_id": 1,
                "data_object": {"$first": "$data_object"}
            }
        }
    ],
);
```

The form of the resulting documents in this case is e.g.:

```js
{
    "_id" : ObjectId("66b4c7a8a2c847e76f4a1619"),
    "biosample_id" : "nmdc:bsm-11-1gzgce32",
    "data_object" : {
        "_id" : ObjectId("65a0688bfa6e7e39b884012f"),
        "id" : "nmdc:dobj-11-72cj2a39",
        "name" : "nmdc_wfmgas-11-dc2b0f84.1_contigs.fna",
        "description" : "Assembly contigs for nmdc:wfmgas-11-dc2b0f84.1",
        "alternative_identifiers" : [

        ],
        "file_size_bytes" : 127171635,
        "md5_checksum" : "090e8612f7bcf18530bb088f79a3cb54",
        "data_object_type" : "Assembly Contigs",
        "url" : "https://data.microbiomedata.org/data/nmdc:omprc-11-avd3n976/nmdc:wfmgas-11-dc2b0f84.1/nmdc_wfmgas-11-dc2b0f84.1_contigs.fna",
        "type" : "nmdc:DataObject"
    }
}
```

**Comment by sujaypatil96** [comment]
Wow, thanks for adding the tests and a prototype of the implementation with `$graphLookup` @dwinston!! i'll take a look at it. Looks super promising ❤️

**Comment by sujaypatil96** [comment]
@dwinston I ran:

```
make up-test
make test-build
docker compose --file docker-compose.test.yml run test -k 'test_find_data_objects_for_study'
```

The test `test_find_data_objects_for_study()` ran successfully, and returned 61 biosamples and data objects, which is the number of records that we expect to be returned. For the `nmdc:sty-11-hdd4bf83` study, the current API endpoint as we have it also returns 61 biosamples and data objects so this works as expected. I believe it's ready to be merged.

**Comment by sujaypatil96** [comment]
Also, I will capture the code for the aggregation pipeline [here](https://github.com/microbiomedata/nmdc-runtime/pull/608#issuecomment-2276000569) on a separate issue, and we can work on benchmarking the two approaches -- Python (implemented in this PR) and Python / Mongo ([here](https://github.com/microbiomedata/nmdc-runtime/pull/608#issuecomment-2276000569)), and ultimately pick the superior approach, and also use that to implement the general rollup endpoint.

File: nmdc_runtime/api/endpoints/find.py [hunk 1]
```diff
@@ -110,51 +111,75 @@ def find_data_objects(
     return find_resources(req, mdb, "data_object_set")
 
 
+def get_classname_from_typecode(doc_id: str) -> str:
```
Comment (eecavanna) [inline nmdc_runtime/api/endpoints/find.py:114]:
```suggestion
def get_classname_from_typecode(doc_id: str) -> str:
    r"""
    Returns the name of the schema class of which an instance could have the specified `id`.
    
    >>> get_classname_from_typecode("nmdc:sty-11-r2h77870")
    'Study'
    """
```

File: nmdc_runtime/api/db/mongo.py [hunk 2]
```diff
@@ -107,3 +110,28 @@ def mongodump_excluded_collections():
         )
     )
     return excluded_collections
+
+
+def mongorestore_collection(mdb, collection_name, bson_file_path):
```
Comment (eecavanna) [inline nmdc_runtime/api/db/mongo.py:115]: I recommend adding a docstring to this function.

```suggestion
def mongorestore_collection(mdb, collection_name, bson_file_path):
    r"""
    Replaces the specified collection with one that reflects the contents of the 
    specified BSON file.
    """
```

File: nmdc_runtime/api/db/mongo.py [hunk 3]
```diff
@@ -107,3 +110,28 @@ def mongodump_excluded_collections():
...
+        data = bson.decode_all(bson_file.read())
+        if data:
+            mdb.drop_collection(collection_name)
+            mdb[collection_name].insert_many(data)
+            print(
+                f"mongorestore_collection: {len(data)} documents into {collection_name} after drop"
+            )
+
+
+def mongorestore_from_dir(mdb, dump_directory, skip_collections=None):
```
Comment (eecavanna) [inline nmdc_runtime/api/db/mongo.py:126]: I recommend adding a docstring to this function.

File: tests/test_api/test_endpoints.py [hunk 4]
```diff
@@ -1,25 +1,83 @@
...
+SCHEMA_COLLECTIONS_MONGODUMP_ARCHIVE_BASENAME = (
+    "nmdc-prod-schema-collections__2024-07-29_20-12-07"
+)
+SCHEMA_COLLECTIONS_MONGODUMP_ARCHIVE_URL = (
+    "https://portal.nersc.gov/cfs/m3408/meta/mongodumps/"
+    f"{SCHEMA_COLLECTIONS_MONGODUMP_ARCHIVE_BASENAME}.tar"
+)  # 84MB. Should be < 100MB.
+
+
+def ensure_local_mongodump():
```
Comment (eecavanna) [inline tests/test_api/test_endpoints.py:41]: I assume the name "ensure X" is being used to mean "ensure X exists". If that is the case, I'd prefer the function name be updated to say that specifically.

File: tests/test_api/test_endpoints.py [hunk 5]
```diff
@@ -1,25 +1,83 @@
...
+from nmdc_runtime.util import REPO_ROOT_DIR, ensure_unique_id_indexes
+from tests.test_util import download_and_extract_tar
+from tests.test_ops.test_ops import op_context as test_op_context
+
+TEST_MONGODUMPS_DIR = REPO_ROOT_DIR.joinpath("tests", "nmdcdb")
+SCHEMA_COLLECTIONS_MONGODUMP_ARCHIVE_BASENAME = (
+    "nmdc-prod-schema-collections__2024-07-29_20-12-07"
+)
+SCHEMA_COLLECTIONS_MONGODUMP_ARCHIVE_URL = (
+    "https://portal.nersc.gov/cfs/m3408/meta/mongodumps/"
```
Comment (eecavanna) [inline tests/test_api/test_endpoints.py:36]: I am wondering whether there is documentation in this folder on the NERSC filesystem, that people there can use to know that the file is used by this code base. Maybe the file could be moved to a directory named `nmdc-runtime-assets` or something.

**Comment by aclum** [comment]
Is this PR going to be ready soon? I would like this updated endpoint to be out for the end of August release because I would like to get IMG switched over to using our API to find data instead of looking directly at the filesystem.

**Comment by sujaypatil96** [comment]
@aclum yes, this PR is actually ready to be merged. I'm addressing some of @eecavanna's review comments now.

File: tests/test_api/test_endpoints.py (hunk 5 above)
Comment (dwinston) [reply tests/test_api/test_endpoints.py:36]: @eecavanna noted this issue. A move seems reasonable, and we should document this. Started microbiomedata/infra-admin#110

**Review by dwinston — APPROVED** [review]

**Pull request merged**